pallets/did/src/lib.rs

```diff
@@ -134,26 +174,290 @@ pub mod pallet {
+				Self::ensure_controller(controller, &did_doc);
+
+				T::Currency::unreserve(&did.clone().into(), T::DidDeposit::get());
+				Self::deposit_event(Event::DidRemoved{
+					did,
+				});
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000000)]
+		pub fn force_remove_did(
+			origin: OriginFor<T>,
+			did: DidIdentifierOf<T>,
+		) -> DispatchResult {
+			// Origin ONLY GovernanceOrigin
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			Did::<T>::try_mutate_exists(did.clone(), |maybe_doc| -> Result<(), DispatchError> {
+				let did_doc = maybe_doc.take().ok_or(Error::<T>::DidNotFound)?;
+				T::Currency::unreserve(&did.clone().into(), T::DidDeposit::get());
+				Self::deposit_event(Event::DidForcedRemoved{
+					did,
+				});
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000000)]
+		pub fn add_did_service(
+			origin: OriginFor<T>,
+			did: DidIdentifierOf<T>,
+			service: Service<T>,
+		) -> DispatchResult {
+			let controller = ensure_signed(origin)?;
+
+			Did::<T>::try_mutate(did.clone(), |maybe_doc| -> Result<(), DispatchError> {
+				let did_doc = maybe_doc.as_mut().ok_or(Error::<T>::DidNotFound)?;
+
+				Self::ensure_controller(controller, &did_doc);
+
+				let service_hash = Self::do_insert_service(service);
+
+				// if services exists, insert new service in sorted location
+				// else insert a new BoundedVec with the new service in the did_doc
+				if let Some(mut services) = did_doc.services.as_mut() {
+					let location = services.binary_search(&service_hash).err().ok_or(Error::<T>::ServiceAlreadyInDid)?;
+					services.try_insert(location, service_hash.clone())
+					.map_err(|_| Error::<T>::TooManyServicesInDid)?;
+				} else {
+					let mut services: BoundedVec<KeyIdOf<T>, T::MaxServices> = BoundedVec::default();
+					// Only fails if `MaxServices` is 0
+					services.try_push(service_hash).map_err(|_| Error::<T>::TooManyServicesInDid)?;
+					did_doc.services = Some(services);
+				};
+
+				Self::deposit_event(Event::DidServiceAdded{
+					did,
+					id: service_hash,
+				});
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000000)]
+		pub fn remove_did_service(
+			origin: OriginFor<T>,
+			did: DidIdentifierOf<T>,
+			service: Service<T>,
+		) -> DispatchResult {
+			let controller = ensure_signed(origin)?;
+
+			Did::<T>::try_mutate(did.clone(), |maybe_doc| -> Result<(), DispatchError> {
+				let did_doc = maybe_doc.as_mut().ok_or(Error::<T>::DidNotFound)?;
+				// ensure that the caller is the controller of the DID
+				Self::ensure_controller(controller, &did_doc);
+
+				let deleted_hash = if let Some(mut services) = did_doc.services.as_mut() {
+					let service_hash = T::Hashing::hash_of(&service);
+					let location = services.binary_search(&service_hash).ok().ok_or(Error::<T>::ServiceNotInDid)?;
+					services.remove(location)
+				} else {
+					return Err(Error::<T>::ServiceNotInDid)?
+				};
+
+				Self::deposit_event(Event::DidServiceRemoved{
+					did,
+					id: deleted_hash
+				});
+				Ok(())
+			})
```

Comment:
```suggestion
			})?;
			Ok(())
```
Preparing for returning a `DispatchResultWithPostInfo` with another Ok(weight) based on the number of `Services`